[DebugInfo] Don't apply is_stmt on MBB branches that preserve lines

SLTozer opened the pull request:
This patch follows on from the changes made in 5fef40c, by adding an additional heuristic that prevents us from applying the start-of-MBB is_stmt flag when we can see that, for all direct branches to the MBB, the last line stepped on before the branch is the same as the first line of the MBB. This is mainly to prevent certain pathological cases, such as macros that expand to multiple basic blocks that all have the same source location, from giving us repeated steps on the same line. This approach is not comprehensive, since it relies on analyzeBranch to read edges, but the default fallback of applying is_stmt may lead only to useless steps in some cases, rather than skipping useful steps altogether.

SLTozer committed:
[DebugInfo] Don't apply is_stmt on MBB branches that preserve lines

This patch follows on from the changes made in 5fef40c, by adding an
additional heuristic that prevents us from applying the start-of-MBB
is_stmt flag when we can see that, for all direct branches to the MBB,
the last line stepped on before the branch is the same as the first line
of the MBB. This is mainly to prevent certain pathological cases, such as
macros that expand to multiple basic blocks that all have the same source
location, from giving us repeated steps on the same line. This approach
is not comprehensive, since it relies on analyzeBranch to read edges.

pogo59 commented:
Stephen wrote this patch at my request; we have code that will now (after the previous patch to set is_stmt on each bb) stop multiple times on the same line, because of a macro that expands to multiple bbs.

Because I asked for it, but it looks more complex than I was hoping, I think other people should review it. Also, can we get some compile-time performance data? It must degrade some.

OCHyams commented:
> [...do this to avoid] giving us repeated steps on the same line [...]

Is there an argument that it might be reasonable to leave this decision to the debugger?

SLTozer committed:
Fixes and simplification

SLTozer commented:
> Is there an argument that it might be reasonable to leave this decision to the debugger?

Possibly, but there might be a justification to do this regardless for the purposes of reducing file size. This patch has a performance cost of 0.17% for `-O0 -g` builds (no notable increase for ReleaseLTO-g) with a self-built clang, which suggests some further optimization is required (likely narrowing down the cases that we'll be willing to handle). However, the original patch gave a 0.33% size increase at `-O0 -g`, and a 0.17% at ReleaseLTO-g, which this patch brings down to 0.01% and 0.04% respectively. I haven't managed to reproduce the change in file size noted by @dwblaikie on the previous PR https://github.com/llvm/llvm-project/pull/105524, but if the effect of this patch is proportional then even the performance cost of this patch could be considered worthwhile.

SLTozer commented:
Updated the patch and added a fix - it has a negligible performance cost for ReleaseLTO-g builds on [the compile time tracker](http://llvm-compile-time-tracker.com/compare.php?from=704116373ae91a1b829dc3d3d269874fb27b579c&to=5830883bd08661a30be542744f6ecdca4c23a09e&stat=instructions%3Au), and combined with the previous patch there is a 0.05% performance cost [when combined with the previous patch](http://llvm-compile-time-tracker.com/compare.php?from=57c64546b616879a28f0efcc60501c6e61670c49&to=5830883bd08661a30be542744f6ecdca4c23a09e&stat=instructions%3Au). It has a larger cost at O0, but it is also necessary for significantly reducing .debug_line inflation, and the original bug was reported at O0 so I'd _personally_ say that the patches collectively are worthwhile. This depends on review somewhat, though.

SLTozer commented:
Small ping for @dwblaikie - I'm preparing to revert the previous commit (some tests that were updated since it landed have to be updated), with the intent to reland the changes as part of this PR. Since I've not been able to reproduce the file size increases that you noted, are you able to test this patch to see if it reduces the impact of the previous patch?

dwblaikie commented:
> Small ping for @dwblaikie - I'm preparing to revert the previous commit (some tests that were updated since it landed have to be updated), with the intent to reland the changes as part of this PR. Since I've not been able to reproduce the file size increases that you noted, are you able to test this patch to see if it reduces the impact of the previous patch?

Hrm :/ what did you try testing?

I'll see if I can find someone to investigate this on our side to get more data...

SLTozer commented:
> Hrm :/ what did you try testing?

Simple stage 2 build of clang, tried Debug and RelWithDebInfo with CMake defaults, with the commit applied and reverted. I didn't do split dwarf, but I don't think that should affect the line table size? If you have any more info on the config I might be able to try it out myself.

cmtice commented:
I tested this PR on the case reported by blaikiew, where we had been seeing  a large regression in the size of .debug_lines.  With this PR we are no longer seeing the size regression.

SLTozer committed:
Minor fixups, comment adjustments

SLTozer commented:
Ping, did a rebase and made some minor fixes.

SLTozer committed:
Re-add unnecessarily removed code

jmorse commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2229,6 +2230,112 @@ void DwarfDebug::beginFunctionImpl(const MachineFunction *MF) {
   PrologEndLoc = emitInitialLocDirective(
       *MF, Asm->OutStreamer->getContext().getDwarfCompileUnitID());
+
+  // For each MBB we try to determine whether and what source line is *always*
+  // the last stepped-on line before entering MBB. Such a line exists if every
```
This is nitpicking, but I feel the order of this sentence should be flipped to make it clear at the start what's going on -- i.e. "Find blocks where the line stepped on before entering it isn't always the same". Otherwise it's quite oblique to work out what the subject is, until the end.

jmorse commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2229,6 +2230,112 @@ void DwarfDebug::beginFunctionImpl(const MachineFunction *MF) {
+  // We only need to examine MBBs that could have is_stmt set by this logic.
+  // We use const_cast even though we won't actually modify MF, because some
+  // methods we need take a non-const MBB.
+  SetVector<MachineBasicBlock *> PredMBBsToExamine;
```
`SmallSetVector` maybe?

jmorse commented on llvm/test/DebugInfo/X86/is_stmt-at-block-start.ll:
```diff
@@ -0,0 +1,62 @@
+;; Checks that when an instruction at the start of a BasicBlock has the same
+;; DebugLoc as the instruction at the end of the previous BasicBlock, we add
+;; is_stmt to the new line, to ensure that we still step on it if we arrive from
+;; a BasicBlock other than the immediately preceding one, unless all known
+;; predecessor BasicBlocks end with the same line.
+
```
Convention is to put the RUN lines above this IIRC? Either way I endorse every test having an explanation of what it's testing.

jmorse reviewed (commented):
This is looking good, although I reckon all the new code should be factored into a new named method so that readers of `beginFunctionImpl` get a better overview of what's going on.

Is there a reason for examining successors of interesting blocks found, instead of examining the predecessor of all blocks? Somehow the latter feels more natural to me; I don't think there's a technical reason for doing it like that though.

I wonder whether there's a topological way of avoiding some of the work, i.e. if we explore in RPO then we might have already found a block that needs is_stmt and can skip analysing it again for other predecessors? That might be overcomplicating it if this doesn't fire very often though.

It pains me to say this but... we must still explicitly skip meta-instructions post-isel to avoid variable location info leaking into the linetables.

SLTozer commented:
> Is there a reason for examining successors of interesting blocks found, instead of examining the predecessor of all blocks? Somehow the latter feels more natural to me; I don't think there's a technical reason for doing it like that though.

The reason we work from the predecessors is that we need to run `analyzeBranch` on the predecessor to get the correct behaviour, and depending on the output of that we may have different outgoing lines to different successors, so it's best to do all the successors for a predecessor at once than to either repeat the check for some MBBs multiple times or to cache the results of analyzeBranch.

We could use RPO, but it would probably be just as or more efficient to just add a slightly-faster early exit at the top of the loop - it's trivial to check for each predecessor whether any of its successors are in the `PotentialIsStmtMBBInstrs` set. In any case, right now the performance cost seems to be pretty low - I think the most important thing is that we never perform unnecessary linear scans, which the earlier-exit should achieve.

SLTozer committed:
Address review comments

jmorse reviewed (approved):
LGTM, I still worry about the clarity of the opening comment (just because there's a lot going on), is there an easy example that can be cooked up to demonstrate this? Fine to land without.

SLTozer committed:
Update comments, remove unnecessary formatted code
